docs/cookbook/recipe_sonata_admin_without_user_bundle.rst

```diff
@@ -0,0 +1,451 @@
+                return false;
+            }
+
+            if (!$user->hasRole('ROLE_ADMIN')) {
+                throw new CustomUserMessageAuthenticationException("You don't have permission to access that page.");
+            }
+
+            return true;
+        }
+
+        public function onAuthenticationFailure(Request $request, AuthenticationException $exception)
+        {
+            $request->getSession()->set(Security::AUTHENTICATION_ERROR, $exception);
+            return new RedirectResponse($this->router->generate('admin_login'));
+        }
+
+        protected function getLoginUrl()
+        {
+            return new RedirectResponse($this->router->generate('admin_login'));
+        }
+
+        public function onAuthenticationSuccess(Request $request, TokenInterface $token, $providerKey)
+        {
+            return new RedirectResponse($this->router->generate('sonata_admin_dashboard'));
+        }
+    }
+
+AdminLoginController
+^^^^^^^^^^^^^^^^^^^^
+
+A Controller, used to render login form. Logout is left empty intentionally because
+this will be handled by Symfony, but we still need to register that route.
+
+.. code-block:: php
+
+    namespace App\Controller;
+
+    use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+    use App\Form\AdminLoginForm;
+    use Symfony\Component\Routing\Annotation\Route;
+    use Symfony\Component\Security\Http\Authentication\AuthenticationUtils;
+
+    class AdminLoginController extends Controller
+    {
+        /**
+         * @var AuthenticationUtils
+         */
+        private $authenticationUtils;
+
+        public function __construct(AuthenticationUtils $authenticationUtils)
+        {
+            $this->authenticationUtils = $authenticationUtils;
+        }
+
+        /**
+         * @Route("/admin/login", name="admin_login")
+         */
+        public function loginAction()
+        {
+            $form = $this->createForm(AdminLoginForm::class, [
+                'email' => $this->authenticationUtils->getLastUsername()
+            ]);
+
+            return $this->render('security/login.html.twig', [
+                'last_username' => $this->authenticationUtils->getLastUsername(),
+                'form' => $form->createView(),
+                'error' => $this->authenticationUtils->getLastAuthenticationError(),
+            ]);
+        }
+
+        /**
+         * @Route("/admin/logout", name="admin_logout")
+         */
+        public function logoutAction()
+        {
+            // Left empty intentionally because this will be handled by Symfony.
+        }
+    }
+
+Setup firewall in `security.yaml`
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. code-block:: yaml
+
+    # config/packages/security.yaml
+
+    security:
+        role_hierarchy:
+            ROLE_ADMIN:       [ROLE_USER, ROLE_SONATA_ADMIN]
+            ROLE_SUPER_ADMIN: [ROLE_ADMIN, ROLE_ALLOWED_TO_SWITCH]
+            SONATA:
+                - ROLE_SONATA_PAGE_ADMIN_PAGE_EDIT
```

Comment:
Hmm, what if SonataPage is not installed/used?

Reply:
Since I copy/pasted that from https://sonata-project.org/bundles/user/4-x/doc/reference/installation.html#integrating-the-bundle-into-the-sonata-admin-bundle
I am not sure it has anything to do with page bundle?

Reply:
I am, but as it is a cookbook article, its fine